Common - Add more loadout randomization options

DartRuffian opened the pull request:
**When merged this pull request will:**
- Add more loadout randomization options
- Respect the [Submitting Content Guidelines](https://github.com/CBATeam/CBA_A3/wiki/Submitting%20content)

DartRuffian committed:
Add more loadout randomization

rautamiekka commented on addons/common/fnc_randomizeLoadout.sqf:
```diff
@@ -0,0 +1,137 @@
+    _unit isKindOf _x || {(vehicleVarName _unit) isEqualTo _x}
+} != -1;
+
+if (_randomizationDisabled || {!(_unit getVariable ["BIS_enableRandomization", true])}) exitWith {true};
+
+if (isNull _unit || !(_unit getVariable ["BIS_enableRandomization", true])) exitWith {};
```
Why check the unit's BIS rando' twice ?

DartRuffian replied:
Originally had this code outside of CBA, forgot to remove the checks from that

DartRuffian committed:
Remove duplicate checks

DartRuffian committed:
Fix issues from when moving to CBA

DartRuffian committed:
Remove (now) unused defines

DartRuffian committed:
Fix loadout array

DartRuffian committed:
Change to Init, fix when slot is empty

DartRuffian committed:
Slight rework

Only add extra mags in InitPost, prevents duplicated mags if mission maker edits loadout.

DartRuffian committed:
Add function to set unit's identity in Eden

DartRuffian commented on addons/common/fnc_fixAnimation3DEN.sqf:
I'm not exactly sure what causes the animation to be wrong (e.g. in the rifle holding anim when the unit only has a pistol). I don't remember ever seeing it before working on this pr though, so I added a workaround for it.

It seems to need to run on the next frame, but CBA's frame functions don't work in Eden so I had to use a sleep with a small delay.

rautamiekka commented on addons/common/fnc_fixAnimation3DEN.sqf:
```diff
@@ -0,0 +1,36 @@
+    sleep 0.1;
+    private _animation = switch (false) do {
+        case (primaryWeapon _this == ""): { "amovpercmstpsraswrfldnon" };
+        case (handgunWeapon _this == ""): { "amovpercmstpsraswpstdnon" };
```
```suggestion
        case (primaryWeapon _this isEqualTo ""): { "amovpercmstpsraswrfldnon" };
        case (handgunWeapon _this isEqualTo ""): { "amovpercmstpsraswpstdnon" };
```

rautamiekka commented on addons/common/fnc_randomizeLoadout.sqf:
```diff
@@ -0,0 +1,133 @@
+
+private _randomizationDisabled = getArray (missionConfigFile >> "disableRandomization") findIf {
+    _unit isKindOf _x || {(vehicleVarName _unit) isEqualTo _x}
+} != -1;
```
```suggestion
} isNotEqualTo -1;
```

rautamiekka commented on addons/common/fnc_randomizeLoadout.sqf:
```diff
@@ -0,0 +1,133 @@
+    private _item = selectRandomWeighted _binocularList;
+    private _magazine = (_item call CBA_fnc_compatibleMagazines) param [0, ""]; // For laser designators
+    _unit addWeapon _item;
+    if (_magazine != "") then {
```
```suggestion
    if (_magazine isNotEqualTo "") then {
```

rautamiekka commented on addons/common/fnc_setIdentity3DEN.sqf:
```diff
@@ -0,0 +1,51 @@
+params ["_unit", ["_identity", ""]];
+TRACE_2("fnc_setIdentity3DEN",_unity,_identity);
+
+if (_identity == "") then {
```
```suggestion
if (_identity isEqualTo "") then {
```

DartRuffian committed:
Update addons/common/XEH_init3DEN.sqf

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

DartRuffian committed:
Fix function names

DartRuffian committed:
Update addons/common/fnc_setIdentity3DEN.sqf

Co-authored-by: Jouni Järvinen <rautamiekka@users.noreply.github.com>

DartRuffian committed:
Fix zeus spawned units, remove function only called once

DartRuffian committed:
Formatting

DartRuffian committed:
Combine ifs

DartRuffian committed:
Fix EH

DartRuffian commented on addons/common/XEH_postInit.sqf:
```diff
@@ -47,3 +47,7 @@ for "_i" from 0 to (count _config - 1) do {
 }] call CBA_fnc_addEventHandler;
+
+// Check missionTime so Eden placed units don't have their loadouts randomized twice
+["CAManBase", "InitPost", { if (CBA_missionTime > 0.1) then { call CBA_fnc_randomizeLoadout } }] call CBA_fnc_addClassEventHandler;
```
The check is so units spawned in Eden don't have their kits randomized again, but units spawned later (e.g. script / zeus) do

DartRuffian replied:
ToDo: Should just be called from CuratorObjectPlaced eh

DartRuffian commented on addons/common/fnc_getRandomizedEquipment.sqf:
```diff
@@ -0,0 +1,54 @@
+            _uniformList, _vestList, _backpackList,
+            _headgearList, _facewearList, _binocularList,
+            _nvgList
+        ] findIf { _x isEqualTo [] } >= -1
```
Ah right, typo, just > should be fine I think

At work currently

DartRuffian committed:
Fix typo

DartRuffian committed:
Delay eden fixAnimation

Add delay to eden fixAnimation call to not change animations when creating editorPreviews.

DartRuffian committed:
Remove animation workaround

DartRuffian committed:
Deleted too much

DartRuffian commented:
Been over a year since this was opened, could some take a look again?

PabstMirror commented on addons/common/XEH_preInit.sqf:
```diff
@@ -74,8 +74,26 @@ activateAddons GVAR(addons);
+// Loadout randomization
+GVAR(randomLoadoutUnits) = createHashMap;
+if (is3DEN) then {
+    add3DENEventHandler ["OnEditableEntityAdded", {
```
this will only work for new missions

take existing mission and `CBA_facewearList[]` gets ignored

because the unit wasn't placed in 3den
and the `InitPost` runs at mission time 0

DartRuffian replied:
I feel like that's probably fine. Could add a special case where units that only have facewear list get randomized at mission start too but I feel like that'd cause more unexpected behavior for people editing kits

PabstMirror commented on addons/common/fnc_randomizeLoadout.sqf:
```diff
@@ -0,0 +1,133 @@
+    // CBA's frame functions don't work in Eden
+    _unit spawn {
+        sleep 0.01;
+        save3DENInventory [get3DENEntityID _this];
```
I don't think we have any way to tell if this has been run,
so it's probably best to just not touch 3den loadout and just run in xeh

DartRuffian replied:
It you randomize at mission start you lose any changes made in efen
